src/snowflake/connector/file_transfer_agent.py

```diff
@@ -380,507 +390,269 @@ def execute(self):
+                notify_file_completed()
+            elif done_client.meta.result_status == ResultStatus.SKIPPED:
+                # this case applies to upload only
+                notify_file_completed()
+            else:
+                logger.debug(f"Finished preparing file {done_client.meta.name}")
+                with cv_chunk_process:
+                    while transfer_metadata.chunks_in_queue > 2 * max_concurrency:
+                        logger.debug(
+                            "Chunk queue busy, waiting in file done callback..."
+                        )
+                        cv_chunk_process.wait()
+                    for _chunk_id in range(done_client.num_of_chunks):
+                        _callback = partial(
+                            transfer_done_cb,
+                            done_client=done_client,
+                            chunk_id=_chunk_id,
+                        )
+                        if is_upload:
+                            network_tpe.submit(
+                                function_and_callback_wrapper,
+                                # Work fn
+                                done_client.upload_chunk,
+                                # Callback fn
+                                _callback,
+                                file_meta,
+                                # Arguments for work fn
+                                _chunk_id,
+                            )
+                        else:
+                            network_tpe.submit(
+                                function_and_callback_wrapper,
+                                # Work fn
+                                done_client.download_chunk,
+                                # Callback fn
+                                _callback,
+                                file_meta,
+                                # Arguments for work fn
+                                _chunk_id,
+                            )
+                        transfer_metadata.chunks_in_queue += 1
+                    cv_chunk_process.notify()
+
+        def transfer_done_cb(
+            success: bool,
+            result: Any,
+            file_meta: "SnowflakeFileMeta",
+            done_client: "SnowflakeStorageClient",
+            chunk_id: int,
+        ):
+            # Note: chunk_id is 0 based while num_of_chunks is count
             logger.debug(
-                f"start {retry_round} round for {qtask.qsize()} items with {thread_number} threads"
+                f"Chunk {chunk_id}/{done_client.num_of_chunks} of file {done_client.meta.name} reached callback"
             )
-            pool = ThreadPoolExecutor(thread_number)
-            triggers = Queue()
-            for _ in range(thread_number):
-                pool.submit(
-                    SnowflakeFileTransferAgent._upload_files_in_queue_thread,
-                    qtask,
-                    self._results,
-                    triggers,
+            with cv_chunk_process:
```

Comment:
try to lock cv_chunk_process here when a chunk is done, but cv_chunk_process in preprocess_done_cb is locked for a big loop, and that preprocess_done_cb is triggered in multi thread manner I am afraid some early done transfer chunks might be blocked here (might not be too long still) until the loop in preprocess_done_cb releases this Condition lock.

Reply:
Let me link the line where this cv is used: https://github.com/snowflakedb/snowflake-connector-python/blob/9ad4b67e21fa4a7919c56d05426c705b443158ca/src/snowflake/connector/file_transfer_agent.py#L447

I think it's safe to say that the code on this spot is non-blocking, and doesn't do much checks a variable and sleeps on the CV. And there is at most `cpu_count` of threads in this cv section. I think it is safe to leave it as is.

I could add extra logging that would indicate that this is an issue. Is that okay with you?